node/full_node_test.go

```diff
@@ -265,59 +265,137 @@ func TestPendingBlocks(t *testing.T) {
+				// during PrepareProposal at height h, vote extensions from previous block (h-1) is available
+				require.Equal([]byte(fmt.Sprintf(expectedExtension, req.Height-1)), extendedCommit.VoteExtension)
+				require.NotNil(extendedCommit.Validator)
+				require.NotNil(extendedCommit.Validator.Address)
+				require.NotEmpty(extendedCommit.ExtensionSignature)
+				ok, err := signingKey.GetPublic().Verify(extendedCommit.VoteExtension, extendedCommit.ExtensionSignature)
+				require.NoError(err)
+				require.True(ok)
+			}
+			return &abci.ResponsePrepareProposal{
+				Txs: req.Txs,
+			}, nil
+		}
+		voteExtension := func(_ context.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+			return &abci.ResponseExtendVote{
+				VoteExtension: []byte(fmt.Sprintf(expectedExtension, req.Height)),
+			}, nil
+		}
+		app.On("Commit", mock.Anything, mock.Anything).Return(&abci.ResponseCommit{}, nil)
+		app.On("PrepareProposal", mock.Anything, mock.Anything).Return(prepareProposalVoteExtChecker)
+		app.On("ProcessProposal", mock.Anything, mock.Anything).Return(&abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil)
+		app.On("FinalizeBlock", mock.Anything, mock.Anything).Return(finalizeBlockResponse)
+		app.On("ExtendVote", mock.Anything, mock.Anything).Return(voteExtension)
+		require.NotNil(app)
+
+		require.NoError(node.Start())
+		require.NoError(waitForAtLeastNBlocks(node, 10, Store))
+		require.NoError(node.Stop())
+		app.AssertExpectations(t)
+	})
+
+	// TestExtendVoteFailure
+	t.Run("TestExtendVoteFailure", func(t *testing.T) {
+		app, node, _ := createNodeAndApp(ctx, voteExtensionEnableHeight, t)
+		require.NotNil(node)
+
+		// Create a channel to signal from extendVoteFailure
+		extendVoteFailureChan := make(chan struct{})
+
+		invalidVoteExtension := func(_ context.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+			extendVoteFailureChan <- struct{}{}
+			return nil, fmt.Errorf("ExtendVote failed")
+		}
+		app.On("ExtendVote", mock.Anything, mock.Anything).Return(invalidVoteExtension)
+
+		// Ensure all necessary methods are mocked
+		app.On("Commit", mock.Anything, mock.Anything).Return(&abci.ResponseCommit{}, nil)
+		app.On("PrepareProposal", mock.Anything, mock.Anything).Return(&abci.ResponsePrepareProposal{Txs: nil}, nil)
+		app.On("ProcessProposal", mock.Anything, mock.Anything).Return(&abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil)
+		app.On("FinalizeBlock", mock.Anything, mock.Anything).Return(finalizeBlockResponse)
+		require.NotNil(app)
+
+		// Start the node
+		require.NoError(node.Start())
+
+		// Wait for blocks to be produced until voteExtensionEnableHeight
+		for i := 1; i < voteExtensionEnableHeight; i++ {
+			select {
+			case <-ctx.Done():
+				require.Fail("timeout waiting for block height")
+			default:
+				time.Sleep(100 * time.Millisecond) // Avoid busy waiting
+			}
+		}
```

Comment:
We have a `waitForAtLeastNBlocks` helper
```suggestion
		require.NoError(waitForAtLeastNBlocks(node, voteExtensionEnableHeight, Store)
```